src/gwproxy/tests/dns_cache.c

```diff
@@ -0,0 +1,569 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (C) 2025  Ammar Faizi <ammarfaizi2@gnuweeb.org>
+ */
+#ifdef NDEBUG
+#undef NDEBUG
+#endif
+#include <stdio.h>
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <unistd.h>
+#include <netdb.h>
+#include <time.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <gwproxy/dns_cache.h>
+
+#ifndef ARRAY_SIZE
+#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof((arr)[0]))
+#endif
+
+static void test_dns_cache_init_free(void)
+{
+	struct gwp_dns_cache *cache;
+	int r;
+
+	/* Test successful initialization */
+	r = gwp_dns_cache_init(&cache, 128);
+	assert(r == 0);
+	assert(cache != NULL);
+
+	/* Test freeing */
+	gwp_dns_cache_free(cache);
+
+	/* Test freeing NULL pointer (should not crash) */
+	gwp_dns_cache_free(NULL);
+
+	/* Test with different bucket sizes */
+	r = gwp_dns_cache_init(&cache, 1);
+	assert(r == 0);
+	assert(cache != NULL);
+	gwp_dns_cache_free(cache);
+
+	r = gwp_dns_cache_init(&cache, 1024);
+	assert(r == 0);
+	assert(cache != NULL);
+	gwp_dns_cache_free(cache);
+
+	printf("test_dns_cache_init_free: passed\n");
+}
+
+static void test_dns_cache_basic_insert_lookup(void)
+{
+	struct gwp_dns_cache *cache;
+	struct gwp_dns_cache_entry *entry;
+	struct addrinfo *ai;
+	struct addrinfo hints;
+	int r;
+
+	r = gwp_dns_cache_init(&cache, 128);
+	assert(r == 0);
+
+	/* Create some test address info */
+	memset(&hints, 0, sizeof(hints));
+	hints.ai_family = AF_INET;
+	hints.ai_socktype = SOCK_STREAM;
+	r = getaddrinfo("127.0.0.1", "80", &hints, &ai);
+	assert(r == 0);
+	assert(ai != NULL);
+
+	/* Insert entry */
+	time_t expire_time = time(NULL) + 300; /* 5 minutes from now */
```

Comment:
Put all variable declarations at the beginning of the scope (so move that `time_t var` to the top). Apply that to other places, too.